### PR TITLE
api: remove [:alpha] protodoc tag and replace with annotation

### PR DESCRIPTION
**Pull request by mattklein123**
This is a first step towards https://github.com/envoyproxy/envoy/issues/17920

A single proto (kafka mesh) has been swapped from using the udpa
file_status annotation to the xds file_status annotation to avoid a
large amount of churn and a forthcoming migration of many alpha/wip
protos to non alpha/wip. The rest will be audited and swapped in
future PRs. This single one was done to make sure the doc machinary
works properly.

Risk Level: None, doc only
Testing: Manual review of docs
Docs Changes: Manually verified
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:507]: cc @alyssawilk

**Comment by phlax** [comment]
docs are rendered here https://storage.googleapis.com/envoy-pr/18218/docs/index.html

**Review by alyssawilk — COMMENTED** [review]
one suggestion, orthogonal of this migration PR (thank you so much for clean up) is we probably should ask that folks tagging things as WIP stick a TODO with their name on it or have a tracking issue so they don't forget it.  Else I suspect many will get lost or forgotten about and git blame may lose track of who to poke about forward progress.

Basically it'd be nice to have tooling like we do for runtime guards to make sure things don't linger in WIP status.  cc @adisuissa

**Comment by mattklein123** [comment]
> one suggestion, orthogonal of this migration PR (thank you so much for clean up) is we probably should ask that folks tagging things as WIP stick a TODO with their name on it or have a tracking issue so they don't forget it. Else I suspect many will get lost or forgotten about and git blame may lose track of who to poke about forward progress.

Yes, this makes sense. I will say that my next PR will start warning on this annotation, which I think will make people much more likely to complain/ask about it for follow up though. :)

**Comment by mattklein123** [comment]
cc @adisuissa @htuch the breaking change detector is calling out my change. Let me know what I should do here? I assume we want to allow this type of change? At least addition?

**Comment by adisuissa** [comment]
> cc @adisuissa @htuch the breaking change detector is calling out my change. Let me know what I should do here? I assume we want to allow this type of change? At least addition?

This is because the new xds wip annotations PR (PR https://github.com/cncf/xds/pull/11) isn't updated in the BSR yet.
@johanbrandhorst Can you please point me to how to sync the xds repo in the BSR (https://buf.build/beta/xds)?

Also opened #18228 because we need to modify our tool to allow breaking  changes for WIP annotations.

**Comment by mattklein123** [comment]
@adisuissa so can I ignore this failure for now?

**Comment by adisuissa** [comment]
> @adisuissa so can I ignore this failure for now?

I think it makes sense, so you won't be blocked by this problem.
One thing to be aware of is that we are going to see the same error in every PR that is rebased after this one, so you may want to disable the api_compat CI target in: https://github.com/envoyproxy/envoy/blob/2e8efdefef86f7749aee3ca30d0752218a07ca8a/.azure-pipelines/pipelines.yml#L247-L248

until the BSR is updated.
That's unfortunate, but I think the only way to avoid such a mismatch in the future (for xds or any other package we depend on) is to use an alternative to buf.

**Comment by mattklein123** [comment]
I don't fully understand what buf does, but once this merges, as long as there are no changes after this, won't it pass?

**Comment by adisuissa** [comment]
buf looks at all the current api protos and their dependencies, and compares against the known package version in the BSR. At the moment the BSR xds package (https://buf.build/beta/xds/history) was last updated on Aug 9th, so your latest PR doesn't doesn't appear there.
Therefore the new annotation will not be recognized, and api_compat will continue to fail, until the BSR xds package is updated.

File: tools/protodoc/protodoc.py
Comment (htuch) [inline tools/protodoc/protodoc.py:197]: I think the original (which I probably added) for alpha is actually wrong. The API status is not the same as the extension implementation status.

File: tools/protodoc/protodoc.py
Comment (htuch) [inline tools/protodoc/protodoc.py:785]: Do we need any protodoc tests for this?

**Review by htuch — COMMENTED** [review]
Generally LGTM, thanks for working on this. A couple of minor comments.

File: tools/protodoc/protodoc.py
Comment (mattklein123) [reply tools/protodoc/protodoc.py:197]: Sorry what do you mean? This I think is just related to the API alpha/wip conversion?

File: tools/protodoc/protodoc.py
Comment (mattklein123) [reply tools/protodoc/protodoc.py:785]: AFAICT there are no protodoc tests. cc @phlax if I am missing them.

I agree having tests would be nice, but I would prefer to not block on that for this change. Can I open a help wanted issue to track?

**Comment by johanbrandhorst** [comment]
> This is because the new xds wip annotations PR (PR [cncf/xds#11](https://github.com/cncf/xds/pull/11)) isn't updated in the BSR yet.
> @johanbrandhorst Can you please point me to how to sync the xds repo in the BSR (https://buf.build/beta/xds)?

Hi again! Re: https://github.com/envoyproxy/envoy/pull/18019#issuecomment-915586576, I'm not surprised 😁. As I explained there, these `beta` organization repositories are not actively maintained by us (Buf) anymore. Originally when launching the BSR beta we added these repositories to make it easy for users to experiment with dependency management, but we've adopted a more long term strategy recently in our first-party support for common libraries and we've narrowed the list to just:

- buf.build/googleapis/googleapis
- buf.build/gogo/protobuf
- buf.build/envoyproxy/protoc-gen-validate

These are automatically synced at 11am ET every day by us, picking up any new changes from the last 24 hours.

The hope is that even these would eventually be adopted by their respective owners, but we are committing to maintaining them even if that does not happen. We're (hopefully understandably) reluctant to add other repositories here, but we recognize that `envoy` is a big open source project with constrained resources, so for this case we'd offer to pick up `xds` as well, since it's probably your most frequently changing dependency. The hope would be that eventually you would pick up that repository management (when/if the BSR proves itself valuable enough to you).

Taking a look at some of [your other dependencies](https://github.com/envoyproxy/envoy/blob/31f48ee7a3c32ed383b46c8940263af9dc9bdcd4/api/buf.yaml#L4-L6), there are a few other deprecated modules:

- buf.build/beta/opencensus
- buf.build/beta/prometheus
- buf.build/beta/opentelemetry

These are probably updated much less frequently and with much less impact to you specifically, so it might not matter to you that they are not being synced anymore, but if it is a concern then _ideally_ the project owners themselves could be convinced to create BSR repositories for them. It may be okay to wait until you notice any issues though.

Please let me know how you'd like to proceed.

**Comment by mattklein123** [comment]
> Let us know how we can help and how you'd like to proceed, we want to make sure buf works well for you.

I still don't fully understand the problem here, but I will defer to @adisuissa on how he wants to proceed. In the meantime pending @htuch comments I will probably just disable the CI job.

File: tools/protodoc/protodoc.py
Comment (phlax) [reply tools/protodoc/protodoc.py:785]: there are currently no tests on landed code.

i have a couple of PRs in the pipeline that would change that

File: tools/protodoc/protodoc.py
Comment (htuch) [reply tools/protodoc/protodoc.py:197]: The WiP warning is:
```
This API is work in progress and is not covered by the :ref:`threat model <arch_overview_threat_model>` 
```

What I'm saying is that even before, with alpha, the confusion of alpha for API and alpha for extension status is confusing. I think these notifications should be driven by extension maturity status not by API. Alpha API probably implies alpha implementation status but not the other way around, so this is under approximating.

File: tools/protodoc/protodoc.py
Comment (htuch) [reply tools/protodoc/protodoc.py:785]: Yeah, I think I was confusing with `protoxform` tests. Opening an issue is fine.

File: tools/protodoc/protodoc.py
Comment (mattklein123) [reply tools/protodoc/protodoc.py:197]: Sorry for being dense but I still don't understand what change you are asking for. I believe this is only triggering for the wip annotation I added? I think extension is handled elsewhere?

Comment (htuch) [reply tools/protodoc/protodoc.py:197]: I'm suggesting it's probably not the right place to be putting in this warning at all.

Comment (mattklein123) [reply tools/protodoc/protodoc.py:197]: This function is called from all the places I need it so it seems fine to me? Can you tell me specifically what you are suggesting I change?

Comment (htuch) [reply tools/protodoc/protodoc.py:197]: Sorry if I'm not being very clear. Here is my position:
* APIs do not have threat model coverage, extensions do.
* Since we are generating these warning based on API, we are missing warning about extensions that have stable APIs but alpha status in their extension implementation status (this information is in a different YAML file / location, CC @phlax). I think we have other warnings that derive from that metadata.

Maybe the wording is just confusing here. If instead we said something like `This API includes features that are currently work-in-progress. Those features marked as work-in-progress in the API are not covered by the Envoy threat model` then we have some value in these additional warning.

Comment (mattklein123) [reply tools/protodoc/protodoc.py:197]: But this applies to core also? I'm really confused how extensions are getting conflated here. There are other warnings for alpha extensions.

I can update the text to make it clear that it's not API wide, but I don't think extension should be included and we also need to include verbiage about breaking changes per the unified text I am now using.

I will update the unified text.

Comment (htuch) [reply tools/protodoc/protodoc.py:197]: I see, good point. Yeah, if you can just make clear it's not API-wide and that it's really a case of implementation status not being finalized until the API is that would be great.

File: tools/protodoc/protodoc.py
Comment (mattklein123) [reply tools/protodoc/protodoc.py:785]: https://github.com/envoyproxy/envoy/issues/18271

**Comment by mattklein123** [comment]
@htuch updated PTAL. In the next change I will actually warn when using any API with one of these annotations. After that I will begin converting some things away from alpha/wip that we know are used in prod.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
